### PR TITLE
Wait no running pods for kube-apiserver are present before scaling down etcd on hibernation

### PR DESCRIPTION
**Pull request by dimityrmirchev**
**How to categorize this PR?**

/area robustness
/kind enhancement

**What this PR does / why we need it**:
In the hibernation flow `etcd` is scaled down before it is ensured that there are no running `kube-apiserver` pods. This PR fixes this. The kapi `hvpa` is now deleted before the scale down of the `kube-apiserver`.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/operation/botanist/controlplane.go
Comment (plkokanov) [inline pkg/operation/botanist/controlplane.go]: Can't we use the `Parallel(...)` function from the flow package here: https://github.com/gardener/gardener/blob/bbe19d1c11c59b7a595decfde227947725f4d7f4/pkg/utils/flow/taskfn.go#L119

File: pkg/client/kubernetes/scaling_test.go
Comment (plkokanov) [inline pkg/client/kubernetes/scaling_test.go]: Can you also add a test for the negative case - where the function times out if there are still running pods?

File: pkg/client/kubernetes/scaling.go
Comment (plkokanov) [inline pkg/client/kubernetes/scaling.go]: nit:
```suggestion

		podList := &corev1.PodList{}
```

**Review by plkokanov — CHANGES_REQUESTED** [review]
Thanks for the PR! I just saw some nits and have a couple questions.

File: pkg/client/kubernetes/scaling.go
Comment (rfranzke) [inline pkg/client/kubernetes/scaling.go]:
```suggestion
		if err := c.List(ctx, podList, client.InNamespace(key.Namespace), client.MatchingLabels(deployment.Spec.Selector.MatchLabels)); err != nil {
```

File: pkg/client/kubernetes/scaling.go
Comment (rfranzke) [inline pkg/client/kubernetes/scaling.go]: Why do we need this? Can't we just check whether there are still `Pod` resources? In this case, we could optimize this function by doing a metadata-only list call.

File: pkg/client/kubernetes/scaling_test.go
Comment (rfranzke) [inline pkg/client/kubernetes/scaling_test.go]: Please do not use mock client for new tests, ref https://github.com/gardener/gardener/blob/master/docs/development/testing.md#writing-unit-tests

File: pkg/operation/botanist/controlplane.go
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]:
```suggestion
		ErrorFormat: utilerrors.NewErrorFormatFuncWithPrefix("failed waiting for deployment scale-down"),
```

File: pkg/client/kubernetes/scaling.go
Comment (rfranzke) [inline pkg/client/kubernetes/scaling.go]:
```suggestion
		if err := c.List(ctx, podList, client.InNamespace(key.Namespace), client.MatchingLabels(deployment.Spec.Selector.MatchLabels), client.Limit(1)); err != nil {
```

**Comment by dimityrmirchev** [comment]
@plkokanov @rfranzke Thank you for the review! I addressed the comments. PTAL

File: pkg/client/kubernetes/scaling.go
Comment (rfranzke) [inline pkg/client/kubernetes/scaling.go]: I just noticed that we already have a generic implementation for this: https://github.com/gardener/gardener/blob/fc1fdfcfb9f029a88005ec3f98178c68d79f49f7/pkg/utils/kubernetes/kubernetes.go#L164-L186

File: pkg/operation/botanist/controlplane.go
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]:
```suggestion
func waitUntilNoPodsExistAnymore(ctx context.Context, c client.Client, namespace string, deploymentNames ...string) error {
```

File: pkg/client/kubernetes/scaling.go
Comment (dimityrmirchev) [reply pkg/client/kubernetes/scaling.go]: This function does not provide a timeout parameter. The function cancels with the passed context. Should we use it or instead leave the new more concrete implementation?

File: pkg/operation/botanist/controlplane.go
Comment (dimityrmirchev) [reply pkg/operation/botanist/controlplane.go]: I renamed the function, but left the last parameter as slice since this is not exported. We can always change it later if there is the use case for it.

File: pkg/client/kubernetes/scaling.go
Comment (rfranzke) [reply pkg/client/kubernetes/scaling.go]: Typically, it is the responsibility of the caller to pass a context with a proper timeout, e.g. like in https://github.com/gardener/gardener/blob/bc1827317538914af9be133cba42dfe7bb806e24/pkg/operation/botanist/component/clusterautoscaler/bootstrap.go#L85-L88

Comment (dimityrmirchev) [reply pkg/client/kubernetes/scaling.go]: I reused the generic function and removed the concrete implementation of the `scaling` package.

**Review by rfranzke — COMMENTED** [review]
Thanks, this looks good now!

/approve
Waiting for final review of @plkokanov

File: pkg/operation/botanist/controlplane.go
Comment (plkokanov) [inline pkg/operation/botanist/controlplane.go]: The creation of the deployment object should be outside of the closure. The way that it is implemented now, if you have 3 deployments, all 3 executions will be done with the 3rd deployment name, since the `deploymentName` variable will be equal to the 3rd deployment name after the loop is finished.

```
	for _, deploymentName := range deployments {
		deployment := &appsv1.Deployment{
			ObjectMeta: metav1.ObjectMeta{
				Name:      deploymentName,
				Namespace: namespace,
			},
		}
		fns = append(fns, func(ctx context.Context) error {
			if err := c.Get(ctx, client.ObjectKeyFromObject(deployment), deployment); err != nil {
				return err
			}

			podList := &metav1.PartialObjectMetadataList{}
			podList.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("PodList"))
			timeoutContext, cancel := context.WithTimeout(ctx, time.Minute*5)
			defer cancel()
			return kutil.WaitUntilResourcesDeleted(timeoutContext, c, podList, time.Second*5, client.InNamespace(namespace), client.MatchingLabels(deployment.Spec.Selector.MatchLabels), client.Limit(1))
		})
	}
```

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/controlplane.go
Comment (dimityrmirchev) [reply pkg/operation/botanist/controlplane.go]: Nice catch! Thanks

**Pull request merged**
